### DISCUSSION TITLE
Issue with LLCC68: Seeking Assistance with Ebyte LoRa Module E220-400M22S Integration with Arduino UNO

### DISCUSSION THREAD

**Discussion by krishank652** [jgromes/RadioLib#944] [Q&A]
Dear Community and Developers,

I am encountering challenges in the development of a wireless transmitter device utilizing the Ebyte LoRa module E220-400M22S, which features the LLCC68 chip. Below is the code snippet I am currently using:

```cpp
#include <RadioLib.h>

// E220-400M22S (LLCC68) has the following connections with Arduino UNO:
// NSS pin:   10
// DIO1 pin:  2
// NRST pin:  3
// BUSY pin:  8
// SCK pin:   13   Default Arduino UNO SPI pin for SCK
// MISO pin:  12   Default Arduino UNO SPI pin for MISO
// MOSI pin:  11   Default Arduino UNO SPI pin for MOSI
// NSS pin:   10   Default Arduino UNO SPI pin for SS

LLCC68 llcc68 = new Module(10, 2, 3, 8);

void setup() {
  Serial.begin(9600);
  int state = llcc68.begin(434.0, 125.0, 2, 7, 0x1424, 10, 8, 2.0, false);
  llcc68.setCurrentLimit(100.0);
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("Success!"));
  } else {
    Serial.print(F("Failed, code "));
  }

  delay(100);
}

void loop() {
  int state = llcc68.transmit("01");
  if (state == RADIOLIB_ERR_NONE) {
    // The packet was successfully transmitted
    Serial.println(F("Success!"));
  } else if (state == RADIOLIB_ERR_PACKET_TOO_LONG) {
    Serial.println(F("Too long!"));
  } else if (state == RADIOLIB_ERR_TX_TIMEOUT) {
    Serial.println(F("Timeout!"));
  } else {
    // Some other error occurred
    Serial.print(F("Failed, code "));
    Serial.println(state);
  }

  delay(500);
}
```

I am encountering the following error in the Serial Monitor:
`12:41:58.228 -> Failed, code -2`

My connections are as follows:

```
E220-400M22S (LLCC68)       :         Arduino UNO
NSS                         :         10
DIO1                       :          2
NRST                       :          3
BUSY                       :          8
SCK                         :         13 
MISO                      :          12  
MOSI                      :          11   
NSS                         :          10   
```

For more details on the Ebyte LoRa E220-400M22S module, please refer to the following links:
- [Ebyte LoRa E220-400M22S Detail Page](https://www.cdebyte.com/products/E220-400M22S)
- [Ebyte LoRa E220-400M22S Manual](https://www.cdebyte.com/pdf-down.aspx?id=930)

1. I am getting Error on Serial Monitor `12:41:58.228 -> Failed, code -2` What is the reason of this?
2. In Ebyte LoRa E220-400M22S Manual required voltage is `Min: 1.8v, Normal: 3.3v, Max: 3.7v `(Is I need to change in code?)
3. Power Consumption as per Manual is: `Tx Current: 100mA, Rx Current: 6.5mA, Sleep Current: 180uA` ( Is I need to change in code?)

I appreciate any guidance or assistance the community can provide in resolving this issue.

Thank you,
Krishan Kumar

**Comment by jgromes** [comment]
To answer your questions:

> 1. I am getting Error on Serial Monitor 12:41:58.228 -> Failed, code -2 What is the reason of this?

The meaning of that error is shown here: https://github.com/jgromes/RadioLib/blob/86136d3b46469f4a16d804c052bc2ca8ac3d62d6/src/TypeDef.h#L114-L118

The library way simply not able to find the LLCC68 transceiver on the SPI itnerface, which is most commonly a hardware issue. Try to run the program again with [debug and verbose mode enabled](https://github.com/jgromes/RadioLib/wiki/Debug-mode) and that should tell you more.

> 2. In Ebyte LoRa E220-400M22S Manual required voltage is Min: 1.8v, Normal: 3.3v, Max: 3.7v (Is I need to change in code?)

That's probably the supply voltage range of the E220 module. Are you powering it from 3.3V? And since Arduino Uno is a device with 5V logic, do you have a logic level shifter? The LLCC68 pins are most likely not 5V-tolerant.

> 3. Power Consumption as per Manual is: Tx Current: 100mA, Rx Current: 6.5mA, Sleep Current: 180uA ( Is I need to change in code?)

That's just the power consumption of the module in different modes, I don't understand what you want to change in code.

**Reply by krishank652** [reply]
Thanks for quick reply 🤠.

As you told me to run the program with debug and verbose mode. I don't understand how can I implement this in my code? (You mean that I need to change these implementation in the library code from Library folder?)

Yes, I am using 3.3v from Arduino board.

Is my previous code and wiring connection are correct or not?

**Reply by jgromes** [reply]
> (You mean that I need to change these implementation in the library code from Library folder?)

The instructions are on the wiki page linked above; for you, it is most likely simplest to just edit `RadioLib/src/BuildOptUser.h`

> Is my previous code and wiring connection are correct or not?

Seems correct from what I can see, apart from the issue with logic levels - you cannot directly connect a 3.3V device (LLCC68) and 5V device (Arduino Uno), you will damage the LLCC68. You need some converter between the two, which oyu did not mention, so my guess is that you don't have any.

**Reply by krishank652** [reply]
I am utilizing a 3.3V supply from the Arduino Board, not 5V. Therefore, I believe there is no need for a converter, and I think it's safe from any risk of burning the LLCC68. Please see the picture I am utilizing 3.3V from Board. Can you confirm if my understanding is accurate?

![image](https://github.com/jgromes/RadioLib/assets/157563241/8022c845-e8f1-4875-982b-a8af50b63db0)


I have run the code with [debug and verbose mode enabled](https://github.com/jgromes/RadioLib/wiki/Debug-mode) I got some extra details.
I Got below While debugging:
```
00:21:42.889 -> RadioLib Debug Info
00:21:42.889 -> Version:  6.4.0.0
00:21:42.889 -> Platform: Arduino AVR
00:21:42.936 -> Compiled: Jan 25 2024 00:15:26
00:21:42.983 -> 
00:21:42.983 -> CMDW  80  
00:21:42.983 -> SI  0 
00:21:42.983 -> SO  55  
00:21:42.983 -> CMDR  C0  
00:21:42.983 -> SI  0 
00:21:42.983 -> SO  0 
00:21:43.030 -> CMDR  1D  3 20  
00:21:43.030 -> SI  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
00:21:43.077 -> SO  0 0 80  0 80  80  0 80  0 80  0 80  15  55  55  55  51  
00:21:43.124 -> SX126x not found! (1 of 10 tries) RADIOLIB_SX126X_REG_VERSION_STRING:
00:21:43.171 -> 0000320 00 80 00 80 80 00 80 00 80 00 80 15 55 55 55 51 | ............UUUQ
00:21:43.265 -> Expected string: LLCC68
00:21:43.311 -> CMDW  80  
00:21:43.311 -> SI  0 
00:21:43.311 -> SO  55  
00:21:43.311 -> CMDR  C0  
00:21:43.311 -> SI  0 
00:21:43.311 -> SO  0 
00:21:43.360 -> CMDR  1D  3 20  
00:21:43.360 -> SI  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
00:21:43.406 -> SO  0 0 80  0 0 0 80  80  0 80  0 80  10  82  8A  8A  8A  
00:21:43.455 -> SX126x not found! (2 of 10 tries) RADIOLIB_SX126X_REG_VERSION_STRING:
00:21:43.502 -> 0000320 00 80 00 00 00 80 80 00 80 00 80 10 82 8a 8a 8a | ................
00:21:43.596 -> Expected string: LLCC68
00:21:43.642 -> CMDW  80  
00:21:43.642 -> SI  0 
00:21:43.642 -> SO  55  
00:21:43.642 -> CMDR  C0  
00:21:43.642 -> SI  0 
00:21:43.642 -> SO  0 
00:21:43.642 -> CMDR  1D  3 20  
00:21:43.689 -> SI  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
00:21:43.737 -> SO  0 0 80  0 0 0 80  80  0 80  0 82  8A  8A  8A  8A  8A  
00:21:43.784 -> SX126x not found! (3 of 10 tries) RADIOLIB_SX126X_REG_VERSION_STRING:
00:21:43.831 -> 0000320 00 80 00 00 00 80 80 00 80 00 82 8a 8a 8a 8a 8a | ................
00:21:43.925 -> Expected string: LLCC68
00:21:43.925 -> CMDW  80  
00:21:43.971 -> SI  0 
00:21:43.971 -> SO  55  
00:21:43.971 -> CMDR  C0  
00:21:43.971 -> SI  0 
00:21:43.971 -> SO  0 
00:21:43.971 -> CMDR  1D  3 20  
00:21:44.019 -> SI  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
00:21:44.065 -> SO  0 0 80  0 0 0 80  80  0 80  0 80  15  55  55  55  55  
00:21:44.112 -> SX126x not found! (4 of 10 tries) RADIOLIB_SX126X_REG_VERSION_STRING:
00:21:44.159 -> 0000320 00 80 00 00 00 80 80 00 80 00 80 15 55 55 55 55 | ............UUUU
00:21:44.253 -> Expected string: LLCC68
00:21:44.253 -> CMDW  80  
00:21:44.300 -> SI  0 
00:21:44.300 -> SO  55  
00:21:44.300 -> CMDR  C0  
00:21:44.300 -> SI  0 
00:21:44.300 -> SO  0 
00:21:44.300 -> CMDR  1D  3 20  
00:21:44.346 -> SI  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
00:21:44.346 -> SO  0 0 80  0 80  0 80  80  0 80  0 80  0 8A  8A  8A  8A  
00:21:44.440 -> SX126x not found! (5 of 10 tries) RADIOLIB_SX126X_REG_VERSION_STRING:
00:21:44.487 -> 0000320 00 80 00 80 00 80 80 00 80 00 80 00 8a 8a 8a 8a | ................
00:21:44.583 -> Expected string: LLCC68
00:21:44.583 -> CMDW  80  
00:21:44.583 -> SI  0 
00:21:44.631 -> SO  55  
00:21:44.631 -> CMDR  C0  
00:21:44.631 -> SI  0 
00:21:44.631 -> SO  0 
00:21:44.631 -> CMDR  1D  3 20  
00:21:44.631 -> SI  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
00:21:44.679 -> SO  0 0 80  0 80  80  0 80  0 80  0 82  8A  8A  8A  8A  8A  
00:21:44.726 -> SX126x not found! (6 of 10 tries) RADIOLIB_SX126X_REG_VERSION_STRING:
00:21:44.820 -> 0000320 00 80 00 80 80 00 80 00 80 00 82 8a 8a 8a 8a 8a | ................
00:21:44.916 -> Expected string: LLCC68
00:21:44.916 -> CMDW  80  
00:21:44.916 -> SI  0 
00:21:44.916 -> SO  55  
00:21:44.963 -> CMDR  C0  
00:21:44.963 -> SI  0 
00:21:44.963 -> SO  0 
00:21:44.963 -> CMDR  1D  3 20  
00:21:44.963 -> SI  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
00:21:45.009 -> SO  0 0 80  0 80  80  0 80  0 80  0 80  0 80  0 82  8A  
00:21:45.056 -> SX126x not found! (7 of 10 tries) RADIOLIB_SX126X_REG_VERSION_STRING:
00:21:45.150 -> 0000320 00 80 00 80 80 00 80 00 80 00 80 00 80 00 82 8a | ................
00:21:45.212 -> Expected string: LLCC68
00:21:45.259 -> CMDW  80  
00:21:45.259 -> SI  0 
00:21:45.259 -> SO  55  
00:21:45.259 -> CMDR  C0  
00:21:45.259 -> SI  0 
00:21:45.259 -> SO  0 
00:21:45.306 -> CMDR  1D  3 20  
00:21:45.306 -> SI  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
00:21:45.353 -> SO  0 0 80  0 0 0 80  80  0 80  0 80  0 8A  8A  8A  8A  
00:21:45.400 -> SX126x not found! (8 of 10 tries) RADIOLIB_SX126X_REG_VERSION_STRING:
00:21:45.463 -> 0000320 00 80 00 00 00 80 80 00 80 00 80 00 8a 8a 8a 8a | ................
00:21:45.558 -> Expected string: LLCC68
00:21:45.558 -> CMDW  80  
00:21:45.558 -> SI  0 
00:21:45.606 -> SO  55  
00:21:45.606 -> CMDR  C0  
00:21:45.606 -> SI  0 
00:21:45.606 -> SO  0 
00:21:45.606 -> CMDR  1D  3 20  
00:21:45.653 -> SI  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
00:21:45.653 -> SO  0 0 80  80  0 80  0 80  0 80  0 80  10  94  55  55  55  
00:21:45.700 -> SX126x not found! (9 of 10 tries) RADIOLIB_SX126X_REG_VERSION_STRING:
00:21:45.794 -> 0000320 00 80 80 00 80 00 80 00 80 00 80 10 94 55 55 55 | .............UUU
00:21:45.888 -> Expected string: LLCC68
00:21:45.888 -> CMDW  80  
00:21:45.888 -> SI  0 
00:21:45.934 -> SO  55  
00:21:45.934 -> CMDR  C0  
00:21:45.934 -> SI  0 
00:21:45.934 -> SO  0 
00:21:45.934 -> CMDR  1D  3 20  
00:21:45.934 -> SI  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
00:21:45.981 -> SO  0 0 80  0 80  80  0 80  0 80  0 80  42  8A  88  8A  8A  
00:21:46.028 -> SX126x not found! (10 of 10 tries) RADIOLIB_SX126X_REG_VERSION_STRING:
00:21:46.121 -> 0000320 00 80 00 80 80 00 80 00 80 00 80 42 8a 88 8a 8a | ...........B....
00:21:46.215 -> Expected string: LLCC68
00:21:46.215 -> No SX126x found!
00:21:46.262 -> CMDW  D 8 E7  
00:21:46.262 -> SI  28  
00:21:46.262 -> SO  28  
00:21:46.262 -> failed, code CMDW 80  
00:21:46.309 -> SI  0 
00:21:46.309 -> SO  0 
00:21:46.356 -> failed, code -2
00:21:56.356 -> CMDW  80  
00:21:56.356 -> SI  0 
00:21:56.356 -> SO  0 
00:21:56.356 -> failed, code -2
00:22:06.353 -> CMDW  80  
00:22:06.353 -> SI  0 
00:22:06.400 -> SO  0 
00:22:06.400 -> failed, code -2
```

**Please Help me to build Communication between Arduino UNO and Ebyte Lora E220-400M22S**

**Reply by jgromes** [reply]
> I am utilizing a 3.3V supply from the Arduino Board, not 5V.

Yes, but the I/O pins (NSS, MISO, MOSI etc.) all have 5V voltage logic. When the pin is logic 1, it gets pulled to 5V. The LLCC68 can only tolerate 3.3V on its IO pins. You **are** burning it.

> I Got below While debugging:

The output from the LLCC68 seems quite erratic. Either the wiring is not very good, or it's damaged already.

**Reply by krishank652** [reply]
ok dear, I understand. Now I am trying to use 3.3V power for MOSI MISO NSS SCK, etc. with new Module. Old may be already burnt. Please don't mark it closed. After successfull I will update here.

**Reply by Cosmic-Aerospace-Technologies** [reply]
Any success? Lol, hopefully after a year it is:) Encountering similar issues.
